testing/trino-product-tests/src/main/java/io/trino/tests/product/utils/DeltaQueryExecutor.java

```diff
@@ -0,0 +1,133 @@
+import io.trino.tempto.query.JdbcQueryExecutor;
+import org.sonatype.aether.artifact.Artifact;
+
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.net.URLClassLoader;
+import java.sql.Connection;
+import java.sql.Driver;
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.util.List;
+import java.util.Properties;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static io.airlift.resolver.ArtifactResolver.MAVEN_CENTRAL_URI;
+import static io.airlift.resolver.ArtifactResolver.USER_LOCAL_REPO;
+import static java.lang.ClassLoader.getPlatformClassLoader;
+
+public class DeltaQueryExecutor
+        extends JdbcQueryExecutor
+{
+    private static final DeltaConnectionsPool CONNECTIONS_POOL = new DeltaConnectionsPool();
+
+    public DeltaQueryExecutor(TestContext testContext)
+    {
+        super(testContext.getDependency(JdbcConnectivityParamsState.class, "delta"), CONNECTIONS_POOL, testContext);
+    }
+
+    private static class DeltaConnectionsPool
+            extends JdbcConnectionsPool
+    {
+        private Connection connection;
+
+        @Override
+        public Connection connectionFor(JdbcConnectivityParamsState jdbcParamsState)
+                throws SQLException
+        {
+            if (connection == null || connection.isClosed()) {
+                Properties properties = new Properties();
+                properties.put("user", jdbcParamsState.user);
+                properties.put("password", jdbcParamsState.password);
+                connection = getDatabaseDriver(jdbcParamsState).connect(jdbcParamsState.url, properties);
+            }
+
+            if (connection == null) {
+                // this should never happen, `javax.sql.DataSource#getConnection()` should not return null
+                throw new IllegalStateException("No connection was created for: " + jdbcParamsState.getName());
+            }
+            if (!jdbcParamsState.prepareStatements.isEmpty()) {
+                try (Statement statement = connection.createStatement()) {
+                    for (String query : jdbcParamsState.prepareStatements) {
+                        try {
+                            statement.execute(query);
+                        }
+                        catch (SQLException e) {
+                            throw new SQLException("Preparatory statement failed: " + query, e);
+                        }
+                    }
+                }
+            }
+            return connection;
+        }
+    }
+
+    private static Driver getDatabaseDriver(JdbcConnectivityParamsState jdbcParamsState)
+    {
+        // TODO Add support for maven coordinate in tempto
+        ArtifactResolver resolver = new ArtifactResolver(USER_LOCAL_REPO, ImmutableList.of(MAVEN_CENTRAL_URI));
+        List<URL> classPath = resolver.resolveArtifacts(getArtifact(jdbcParamsState.driverClass)).stream()
```

Comment:
i find this approach very cool!

Note that this is going to interact with maven repositories (eg maven central), so is potential source of flakiness.
can we somehow make sure that artifacts are pre-downloaded, or maybe we need retries here?

Reply:
`ArtifactResolver` can resolve artifacts without connection to maven central if it already exists in the local repository (`~/.m2/repository` in this case). I agree with retries, but the condition is unclear to me until we actually face the issue.

Reply:
We can retry blindly, if we cannot yet distinguish recoverable and unrecoverable errors